test/acl.test.js

```diff
@@ -472,3 +475,138 @@ describe('access check', function() {
+  var models = {};
+
+  beforeEach(setupAppAndRequest);
+
+  it('contains all authorized roles for a principal if query is allowed', function() {
+    return createACLs('MyTestModel', [
+      {permission: ACL.ALLOW, principalId: '$everyone'},
+      {permission: ACL.ALLOW, principalId: '$authenticated'},
+      {permission: ACL.ALLOW, principalId: 'myRole'},
+    ])
+    .then(function() {
+      return request.get('/MyTestModels')
+        .set('X-Access-Token', accessToken.id)
+        .expect(200);
+    })
+    .then(function(res) {
+      return models.MyTestModel.onBeforeRemote;
+    })
+    .then(function(ctx) {
+      expect(ctx.args.options.authorizedRoles).to.eql(
+        ['$everyone', '$authenticated', 'myRole']
+      );
+    });
+  });
+
+  it('does not contain any denied role even if query is allowed', function() {
+    return createACLs('MyTestModel', [
+      {permission: ACL.ALLOW, principalId: '$everyone'},
+      {permission: ACL.DENY,  principalId: '$authenticated'},
+      {permission: ACL.ALLOW, principalId: 'myRole'},
+    ])
+    .then(function() {
+      return request.get('/MyTestModels')
+        .set('X-Access-Token', accessToken.id)
+        .expect(200);
+    })
+    .then(function(res) {
+      return models.MyTestModel.onBeforeRemote;
+    })
+    .then(function(ctx) {
+      expect(ctx.args.options.authorizedRoles).to.eql(
+        ['$everyone', 'myRole']
+      );
+    });
+  });
+
+  it('honors default permission setting', function() {
+    // set MyTestModel default permission as DENY
+    models.MyTestModel.settings.defaultPermission = ACL.DENY;
+
+    return createACLs('MyTestModel', [
+      {permission: ACL.DEFAULT, principalId: '$everyone'},
+      {permission: ACL.DENY,    principalId: '$authenticated'},
+      {permission: ACL.ALLOW,   principalId: 'myRole'},
+    ])
+    .then(function() {
+      return request.get('/MyTestModels')
+        .set('X-Access-Token', accessToken.id)
+        .expect(200);
+    })
+    .then(function(res) {
+      return models.MyTestModel.onBeforeRemote;
+    })
+    .then(function(ctx) {
+      expect(ctx.args.options.authorizedRoles).to.eql(
+        // '$everyone' is not expected as default permission is DENY
+        ['myRole']
+      );
+    });
+  });
+
+  // helpers
+  function setupAppAndRequest() {
+    app = loopback({localRegistry: true, loadBuiltinModels: true});
+    app.use(loopback.rest());
+    app.set('remoting', {errorHandler: {debug: true, log: true}});
+    app.enableAuth();
+    app.dataSource('db', {connector: 'memory'});
+    request = supertest(app);
+
+    // attaching base models
+    ['User', 'Role', 'RoleMapping', 'ACL', 'AccessToken'].map(setupModel);
+
+    // creating a custom model
+    models.MyTestModel = app.registry.createModel('MyTestModel');
+    app.model(models.MyTestModel, {dataSource: 'db'});
+
+    // beforeRemote hook on MyTestModel, as a promise
+    models.MyTestModel.onBeforeRemote = new Promise(function(resolve, reject) {
```

Comment:
Uff, I find this very confusing. Can you simplify to the following?

```js
models.MyTestModel.beforeRemote('find', function(ctx, notUsed,next) {
  models.MyTestModel.lastRemotingContext = ctx;
  next();
});
```

Unless I am missing something here?